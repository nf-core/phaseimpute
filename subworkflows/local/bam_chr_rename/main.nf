include { GAWK                        } from '../../../modules/nf-core/gawk'
include { SAMTOOLS_REHEADER           } from '../../../modules/nf-core/samtools/reheader'
include { SAMTOOLS_INDEX              } from '../../../modules/nf-core/samtools/index'

workflow BAM_CHR_RENAME {
    take:
    ch_bam          // channel: [ [id], bam, index ]
    ch_fasta        // channel: [ [id], fasta, fai ]

    main:

    ch_versions = Channel.empty()
    ch_fasta.view()
    // Generate the chromosome renaming file
    GAWK(ch_fasta.map{ metaG, fasta, fai -> [metaG, fai] }, [])
    ch_versions = ch_versions.mix(GAWK.out.versions)

    cmd_chr = GAWK.out.output
        .splitCsv()
        .map{
            it[1] == "chr" ?
                'awk  \"{ gsub(/^(@SQ.*)(\\tSN:)([0-9XYMT]+)(\\s|\$)/, "\\1chr\\2\\3\\4"); print }\"' :
                'awk  \"{ print gensub(/^(@SQ.*)(\\tSN:)chr/, \"\\\\1\\\\2\", "g"); }\"'
        }
    // Rename the chromosome without prefix
    SAMTOOLS_REHEADER(
        ch_bam, // channel: [ [id], bam, index ]
        cmd_chr
    )
    ch_versions = ch_versions.mix(SAMTOOLS_REHEADER.out.versions.first())

    SAMTOOLS_INDEX(SAMTOOLS_REHEADER.out.bam)
    ch_versions = ch_versions.mix(SAMTOOLS_INDEX.out.versions.first())

    ch_bam_renamed = SAMTOOLS_REHEADER.out.bam
        .combine(SAMTOOLS_INDEX.out.csi, by:0)

    emit:
    bam_renamed    = ch_bam_renamed        // [ [id], bam, csi ]
    versions       = ch_versions           // channel: [ versions.yml ]
}
