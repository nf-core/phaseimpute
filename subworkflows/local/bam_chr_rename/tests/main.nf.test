nextflow_workflow {

    name "Test Subworkflow BAM_CHR_RENAME"
    script "../main.nf"

    config "./nextflow.config"

    workflow "BAM_CHR_RENAME"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/bam_chr_rename"
    tag "bam_chr_rename"

    tag "samtools"
    tag "samtools/reheader"
    tag "samtools/index"
    tag "gawk"

    test("Should run without error") {
        when {
            workflow {
                """
                input[0] = Channel.fromList([
                    [
                        [id: "Paired"],
                        file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam'], checkIfExist:true),
                        file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam_bai'], checkIfExist:true)
                    ],
                    [
                        [id: "MT"],
                        file(params.test_data['homo_sapiens']['illumina']['test_illumina_mt_bam'], checkIfExist:true),
                        file(params.test_data['homo_sapiens']['illumina']['test_illumina_mt_bam_bai'], checkIfExist:true)
                    ]
                ])
                input[1] = Channel.of([
                    [id:"GRCh37_chr22"],
                    file(params.test_data['homo_sapiens']['genome']['genome_fasta'], checkIfExist:true),
                    file(params.test_data['homo_sapiens']['genome']['genome_fasta_fai'], checkIfExist:true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
