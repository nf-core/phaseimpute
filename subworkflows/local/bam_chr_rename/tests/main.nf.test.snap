{
    "Should run without error": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    "versions.yml:md5,41946978e04f660e488582ff0fe9e172",
                    "versions.yml:md5,a05ff29cc17a89640051ed05e53ff0ba"
                ],
                "bam_renamed": [
                    
                ],
                "versions": [
                    "versions.yml:md5,41946978e04f660e488582ff0fe9e172",
                    "versions.yml:md5,a05ff29cc17a89640051ed05e53ff0ba"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-06-16T22:53:49.883716058"
    }
}